Code/GraphMol/ScaffoldNetwork/ScaffoldNetwork.h

```diff
@@ -0,0 +1,177 @@
+#include <RDGeneral/BoostEndInclude.h>
+#endif
+
+namespace RDKit {
+class ROMol;
+class ChemicalReaction;
+
+namespace ScaffoldNetwork {
+
+struct RDKIT_SCAFFOLDNETWORK_EXPORT ScaffoldNetworkParams {
+  bool includeGenericScaffolds =
+      true;  ///< include scaffolds with all atoms replaced by dummies
+  bool includeGenericBondScaffolds =
+      false;  ///< include scaffolds with all bonds replaced by single bonds
+  bool includeScaffoldsWithoutAttachments =
+      true;  ///< remove attachment points from scaffolds and include the result
+  bool includeScaffoldsWithAttachments =
+      true;  ///< Include the version of the scaffold with attachment points
+  bool keepOnlyFirstFragment =
+      true;  ///<  keep only the first fragment from the bond breaking rule
+  bool pruneBeforeFragmenting =
+      true;  ///<  Do a pruning/flattening step before starting fragmenting
+  bool flattenIsotopes = true;  ///< remove isotopes when flattening
+  bool flattenChirality =
+      true;  ///< remove chirality and bond stereo when flattening
+  bool flattenKeepLargest =
+      true;  ///< keep only the largest fragment when doing flattening
+
+  std::vector<std::shared_ptr<ChemicalReaction>>
+      bondBreakersRxns;  ///< the reaction(s) used to fragment. Should expect a
+                         ///< single reactant and produce two products
+  ScaffoldNetworkParams()
+      : ScaffoldNetworkParams{
+            {"[!#0;R:1]-!@[!#0:2]>>[*:1]-[#0].[#0]-[*:2]"}} {};
+  ScaffoldNetworkParams(const std::vector<std::string> &bondBreakersSmarts);
+};
+
+enum class RDKIT_SCAFFOLDNETWORK_EXPORT EdgeType {
+  Fragment = 1,     ///< molecule -> fragment
+  Generic = 2,      ///< molecule -> generic molecule (all atoms are dummies)
+  GenericBond = 3,  ///< molecule -> generic bond molecule (all bonds single)
+  RemoveAttachment = 4,  ///< molecule -> molecule with no attachment points
+  Initialize = 5         ///< molecule -> flattened molecule
+};
+
+struct RDKIT_SCAFFOLDNETWORK_EXPORT NetworkEdge {
+  size_t beginIdx;
+  size_t endIdx;
+  EdgeType type;
+  NetworkEdge() : beginIdx(0), endIdx(0), type(EdgeType::Initialize){};
+  NetworkEdge(size_t bi, size_t ei, EdgeType typ)
+      : beginIdx(bi), endIdx(ei), type(typ){};
+  bool operator==(const RDKit::ScaffoldNetwork::NetworkEdge &o) const {
+    return (beginIdx == o.beginIdx) && (endIdx == o.endIdx) && (type == o.type);
+  }
+  bool operator!=(const RDKit::ScaffoldNetwork::NetworkEdge &o) const {
+    return (beginIdx != o.beginIdx) || (endIdx != o.endIdx) || (type != o.type);
+  }
+#ifdef RDK_USE_BOOST_SERIALIZATION
+ private:
+  friend class boost::serialization::access;
+  template <class Archive>
+  void serialize(Archive &ar, const unsigned int version) {
+    RDUNUSED_PARAM(version);
+    ar &beginIdx;
+    ar &endIdx;
+    ar &type;
+  }
+#endif
+};
+
+struct RDKIT_SCAFFOLDNETWORK_EXPORT ScaffoldNetwork {
+  std::vector<std::string> nodes;  ///< SMILES for the scaffolds
```

Comment:
In my home-grown implementation I stored the smiles and the molecule.  The molecule was used as a query molecule to "apply" the scaffold network to a new set of molecules to get statistics/counts etc.  I wonder if we might make this a NetworkNode type to enable this functionality.  Perhaps:

```
class NetWorkNodeBase {
   const std::string &getHash() = 0;
   bool HasMatch(const ROMol& m) = 0;
}
```

This might already be covered, but I've found it of good utility when applying the scaffold tree to analyze datasets.

Reply:
I can see how it'd be cool to have a tool for doing analysis later, but the idea here is really just to construct and store the network in a really minimal form. 

This also helps with memory consumption when building a network - having to store all the molecules for a large network would be pretty hairy.

Reply:
I think this would be application dependent.  I.e. we just store the smiles and use that for the matching code.  I think it depends on what we do wanting to go forward.  I'll think about it a bit.

Reply:
Agreed that it's app dependent. The classes that are currently there are not ideal for something where you want to do use the scaffold network in an active way. It's really focused on building the network instead of using it.

Reply:
If you're convinced that the string rep is enough, I'm happy with this.  We could always add a smart pointer to hold a molecule at a later date.